### PR TITLE
Fix Interceptor compiler plugin bugs

### PR DESCRIPTION
**Pull request by TharmiganK**
## Purpose
$Subject

> Related to : [`Add compiler validation for interceptors #732`](https://github.com/ballerina-platform/module-ballerina-http/pull/732)

## Examples
N/A

## Checklist
- [ ] <s>Linked to an issue</s>
- [ ] <s>Updated the changelog</s>
- [x] Added tests

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java [hunk 1]
```diff
@@ -61,36 +67,53 @@ public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
...
-                            break;
-                        } else {
-                            reportMultipleReferencesFound(ctx, (TypeReferenceNode) member);
-                            return;
-                        }
+                Optional<Symbol> optionalTypeReferenceSymbol = ctx.semanticModel().symbol(member);
+                if (optionalTypeReferenceSymbol.isPresent()) {
+                    TypeReferenceTypeSymbol typeReferenceSymbol = (TypeReferenceTypeSymbol)
+                            optionalTypeReferenceSymbol.get();
+                    populateTypeReferenceName(ctx, member, typeReferenceSymbol);
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:77]: `populateTypeReferenceName` is returning a `boolean`, but it is omitted here. Is there any specific reason to do this ?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java [hunk 2]
```diff
@@ -61,36 +67,53 @@ public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
...
         }
     }
 
+    private boolean populateTypeReferenceName(SyntaxNodeAnalysisContext ctx, Node node,
+                                              TypeReferenceTypeSymbol typeReferenceSymbol) {
+        List<TypeSymbol> typeInclusions = ((ObjectTypeSymbol) typeReferenceSymbol.typeDescriptor()).typeInclusions();
+
+        if (typeInclusions.isEmpty()) {
+            String typeName = typeReferenceSymbol.getName().isPresent() ? typeReferenceSymbol.getName().get() : null;
+            if (typeName != null) {
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:93]: Let's do an early return here by checking the inverse condition
if (typeName == null) {
    return ...;
}

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java [hunk 3]
```diff
@@ -61,36 +67,53 @@ public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
...
                 }
             }
         }
         if (interceptorType != null) {
             validateInterceptorResourceMethod(ctx, members, interceptorType);
+            interceptorType = null;
         }
     }
 
+    private boolean populateTypeReferenceName(SyntaxNodeAnalysisContext ctx, Node node,
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:87]: What is the usage of returning a `boolean` value from here ?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java [hunk 4]
```diff
@@ -61,36 +67,53 @@ public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
...
-                        }
+                Optional<Symbol> optionalTypeReferenceSymbol = ctx.semanticModel().symbol(member);
+                if (optionalTypeReferenceSymbol.isPresent()) {
+                    TypeReferenceTypeSymbol typeReferenceSymbol = (TypeReferenceTypeSymbol)
+                            optionalTypeReferenceSymbol.get();
+                    populateTypeReferenceName(ctx, member, typeReferenceSymbol);
                 }
             }
         }
         if (interceptorType != null) {
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:81]: For null-checks lets use java util methods.
```suggestion
        if (Objects.nonNull(interceptorType)) {
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java (hunk 3 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:87]: The `populateTypeReferenceName` will traverse through the type inclusion tree to find whether it contains `http:RequestContext` or `http:RequestErrorContext`. If we find both of them as type inclusions, we report an error. And this boolean value is used to stop the tree traversal check when we report the error.

Comment (ayeshLK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:88]: Shall extract this validation logic to another method and use it then. `populateTypeReferenceName` and returning a boolean does not make sense IMO.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java (hunk 4 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java]: Fixed in [`46596b5 `](https://github.com/ballerina-platform/module-ballerina-http/pull/739/commits/46596b5bf7c18942201cf955f8b0ee0aa85776e2)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java (hunk 2 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java]: Fixed in [`46596b5 `](https://github.com/ballerina-platform/module-ballerina-http/pull/739/commits/46596b5bf7c18942201cf955f8b0ee0aa85776e2)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java (hunk 1 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:79]: Removed returning boolean in [`08f5f55`](https://github.com/ballerina-platform/module-ballerina-http/pull/739/commits/08f5f552a95b90458d20b1da3777080ddad926cc)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java (hunk 3 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java]: Fixed in [`08f5f55`](https://github.com/ballerina-platform/module-ballerina-http/pull/739/commits/08f5f552a95b90458d20b1da3777080ddad926cc)

**Review by chamil321 — APPROVED** [review]

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java (hunk 3 above)
Comment (ayeshLK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java]: @TharmiganK as discussed lets do this in the initial way.

**Comment by TharmiganK** [comment]
This PR was created to add some additional compiler validations for interceptors. Since, the bug is very rare and the approach is not consistent with other validations, it's fine to close it.

**Pull request closed without merging**